### ISSUE TITLE
ConvertRGBToHSL creates black output sometimes if red, green, or blue is > 1.0

### ISSUE THREAD

**Issue by njdoyle** [ImageMagick/ImageMagick#1065]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am using the latest version of ImageMagick
- [x] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
I encountered this issue using the `-modulate` option for `convert`. Sometimes, black pixels are erroneously output if the some of the input RGB values are greater than 1.0. The real world example of this is the ringing artifacts introduced by resizing. On edges with white, ringing is introduced that can bump R, G, or B above 1.0. This has tricked the saturation calculation in `ConvertRGBToHSL` in `MagickCore/gem.c` in to dividing by zero in the functions very last else case. This results in occasional scattered black pixels that should probably be white. A workaround I have tried for this is to clamp the input to `ConvertRGBToHSL` but I'm not convinced this is the most correct solution.

### Steps to Reproduce
Compare the output of these two commands:
```
convert logo: -quality 50 jpeg:- | convert - -resize 400x400 logo.good.png
```
```
convert logo: -quality 50 jpeg:- | convert - -resize 400x400 -modulate 100,100,100 logo.bad.png
```

The `-modulate` is effectively a NOOP but shows an instance of the black pixel artifacts at 57:38. This artifact is not present in the command that does not use `-modulate`. 

I've attached example output from the two commands.

### System Configuration

- ImageMagick version: 7.0.7-28 Q16 x86_64
- Environment (Operating system, version and so on): MacOS 10.13.3
- Additional information: libjpeg-turbo 1.5.3

```
$ magick -list configure

Path: /Users/nidoyle/Prefix/lib/ImageMagick-7.0.7//config-Q16HDRI/configure.xml

Name           Value
-------------------------------------------------------------------------------
CC             gcc
CFLAGS         -I/Users/nidoyle/Prefix/include/libxml2 -I/Users/nidoyle/Prefix/include -I/Users/nidoyle/Prefix/include -I/Users/nidoyle/Prefix/include/libpng16 -I/Users/nidoyle/Prefix/include -I/Users/nidoyle/Prefix/include/openjpeg-2.3 -I/Users/nidoyle/Prefix/include -I/Users/nidoyle/Prefix/include/freetype2 -I/Users/nidoyle/Prefix/include/libpng16 -I/Users/nidoyle/Prefix/include -I/Users/nidoyle/Prefix/include   -g -O2 -Wall -mtune=native -fexceptions -D_FORTIFY_SOURCE=0 -Qunused-arguments -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
CODER_PATH     /Users/nidoyle/Prefix/lib/ImageMagick-7.0.7/modules-Q16HDRI/coders
CONFIGURE      ./configure  '--prefix=/Users/nidoyle/Prefix' '--with-gcc-arch=native' '--disable-openmp' 'LDFLAGS=-L/Users/nidoyle/Prefix/lib -Wl,-rpath,/Users/nidoyle/Prefix/lib ' 'CPPFLAGS=-I/Users/nidoyle/Prefix/include '
CONFIGURE_PATH /Users/nidoyle/Prefix/etc/ImageMagick-7/
COPYRIGHT      Copyright (C) 1999-2018 ImageMagick Studio LLC
CPPFLAGS       -I/Users/nidoyle/Prefix/include/ImageMagick-7
CXX            g++
CXXFLAGS       -g -O2 -Qunused-arguments -pthread
DEFS           -DHAVE_CONFIG_H
DELEGATES      bzlib mpeg freetype jng jpeg lcms openjp2 png tiff webp x xml zlib
DISTCHECK_CONFIG_FLAGS 'CPPFLAGS=-I/Users/nidoyle/Prefix/include -I/Users/nidoyle/Prefix/include ' 'LDFLAGS=-L/Users/nidoyle/Prefix/lib -Wl,-rpath,/Users/nidoyle/Prefix/lib -L/Users/nidoyle/Prefix/lib -Wl,-rpath,/Users/nidoyle/Prefix/lib '  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-gslib=no  --with-fontpath=  --with-rsvg=no  --with-perl=no
DOCUMENTATION_PATH /Users/nidoyle/Prefix/share/doc/ImageMagick-7
EXEC-PREFIX    /Users/nidoyle/Prefix
EXECUTABLE_PATH /Users/nidoyle/Prefix/bin
FEATURES       DPC HDRI Cipher
FILTER_PATH    /Users/nidoyle/Prefix/lib/ImageMagick-7.0.7/modules-Q16HDRI/filters
GIT_REVISION   13919
HOST           x86_64-apple-darwin17.4.0
INCLUDE_PATH   /Users/nidoyle/Prefix/include/ImageMagick-7
LDFLAGS        -L/Users/nidoyle/Prefix/lib -L/Users/nidoyle/Prefix/lib -Wl,-rpath,/Users/nidoyle/Prefix/lib -L/Users/nidoyle/Prefix/lib -Wl,-rpath,/Users/nidoyle/Prefix/lib   -L/usr/X11/lib
LIB_VERSION    0x707
LIB_VERSION_NUMBER 7,0,7,28
LIBRARY_PATH   /Users/nidoyle/Prefix/lib/ImageMagick-7.0.7
LIBS             -L/Users/nidoyle/Prefix/lib -llcms2 -ltiff -L/Users/nidoyle/Prefix/lib -lfreetype  -ljpeg   -L/Users/nidoyle/Prefix/lib -lpng16 -lz        -L/Users/nidoyle/Prefix/lib -lwebpmux -lwebp -L/Users/nidoyle/Prefix/lib -lwebp   -lXext -lXt   -lSM -lICE -lX11   -lbz2  -L/Users/nidoyle/Prefix/lib -lopenjp2    -L/Users/nidoyle/Prefix/lib -lxml2  -L/Users/nidoyle/Prefix/lib -lz   -lm
NAME           ImageMagick
PCFLAGS        -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
PREFIX         /Users/nidoyle/Prefix
QuantumDepth   16
RELEASE_DATE   2018-03-27
SHARE_PATH     /Users/nidoyle/Prefix/share/ImageMagick-7
SHAREARCH_PATH /Users/nidoyle/Prefix/lib/ImageMagick-7.0.7/config-Q16HDRI
TARGET_CPU     x86_64
TARGET_OS      darwin17.4.0
TARGET_VENDOR  apple
VERSION        7.0.7
WEBSITE        http://www.imagemagick.org

Path: [built-in]

Name           Value
-------------------------------------------------------------------------------
FEATURES
NAME           ImageMagick
QuantumDepth   16
```
![logo good](https://user-images.githubusercontent.com/316852/37992308-593098ac-31d9-11e8-8b75-81fd6b172885.png)
![logo bad](https://user-images.githubusercontent.com/316852/37992315-5f83ffb4-31d9-11e8-8ef2-5e2091336387.png)

**Comment by fmw42** [comment]
see https://www.imagemagick.org/discourse-server/viewtopic.php?f=3&t=33728

**Comment by njdoyle** [comment]
Looking at commit 4e1e0e892e22510f9fe7740c0ec1ea089b15ac8a, that is separate from the problem I was observing. That commit changes `ConvertHSLToRGB` whereas I was seeing divide by zero in `ConvertRGBToHSL`.

**Comment by urban-warrior** [comment]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in GIT master branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://www.imagemagick.org/download/beta/ by sometime tomorrow.

**Comment by njdoyle** [comment]
This looks like it's fixed in 7.0.7-29. Thank you!
